File: hw2/single_test.m
% just to test one size for all algorithms

i = 3;
n_values = 1:i;
n = n_values(i);
A = hilb(n); % Generate the Hilbert matrix

% Print the matrix
disp('Matrix A:');
disp(A);

% Sherman's March
tic;
[L1, U1] = shermans(A);
time_sherman = toc;
rel_error_sherman = norm(A - L1 * U1, 2) / norm(A, 2);

% Print L and U
disp('Sherman L:');
disp(L1);
disp('Sherman U:');
disp(U1);
disp('sherman error: ');
disp(rel_error_sherman);

% Pickett's Charge
tic;
[L2, U2] = picketts(A);
time_pickett = toc;
rel_error_pickett = norm(A - L2 * U2, 2) / norm(A, 2);

disp('Pickett L:');
disp(L2);
disp('Pickett U:');
disp(U2);
disp('pickett error: ');
disp(rel_error_pickett);

% Crout's Method
tic;
[L3, U3] = crouts(A);
time_crout = toc;
rel_error_crout = norm(A - L3 * U3, 2) / norm(A, 2);

disp('Crout L:');
disp(L3);
disp('Crout U:');
disp(U3);
disp('crout error: ');
disp(rel_error_crout);

% Built-in LU
tic;
[L4, U4] = lu(A);
time_builtin = toc;
rel_error_builtin = norm(A - L4 * U4, 2) / norm(A, 2);

disp('Built-in L:');
disp(L4);
disp('Built-in U:');
disp(U4);
disp('built-in error: ');
disp(rel_error_builtin);

File: hw2/single_test.py
import numpy as np
import matplotlib.pyplot as plt
from time import time
from scipy.linalg import lu, hilbert
from shermans import shermans
from picketts import picketts
from crouts import crouts

# Test size
n = 4
A = hilbert(n)  # Generate Hilbert matrix

print("Matrix A:")
print(A)
print()

# # Sherman's March
# start = time()
# L1, U1 = shermans(A)
# time_sherman = time() - start
# rel_error_sherman = np.linalg.norm(A - L1 @ U1, 2) / np.linalg.norm(A, 2)

# print("Sherman L:")
# print(L1)
# print("Sherman U:")
# print(U1)
# print(f"Sherman error: {rel_error_sherman}\n")
# print(f"Multiplied L1 and U1:")
# print(L1 @ U1)

# # Pickett's Charge
# start = time()
# L2, U2 = picketts(A)
# time_pickett = time() - start
# rel_error_pickett = np.linalg.norm(A - L2 @ U2, 2) / np.linalg.norm(A, 2)

# print("Pickett L:")
# print(L2)
# print("Pickett U:")
# print(U2)
# print(f"Pickett error: {rel_error_pickett}\n")
# print(f"Multiplied L2 and U2:")
# print(L2 @ U2)

# Crout's Method
start = time()
L3, U3 = crouts(A)
time_crout = time() - start
rel_error_crout = np.linalg.norm(A - L3 @ U3, 2) / np.linalg.norm(A, 2)

print("Crout L:")
print(L3)
print("Crout U:")
print(U3)
print(f"Crout error: {rel_error_crout}\n")
print(f"Multiplied L3 and U3:")
print(L3 @ U3)

# Built-in LU
start = time()
P, L4, U4 = lu(A)
time_builtin = time() - start
rel_error_builtin = np.linalg.norm(A - P @ L4 @ U4, 2) / np.linalg.norm(A, 2)

print("Built-in L:")
print(L4)
print("Built-in U:")
print(U4)
print(f"Multiplied L4 and U4:")
print(P @ L4 @ U4)
print(f"Built-in error: {rel_error_builtin}\n")
